### PR TITLE
Fix warnings detected by Visual Studio, Part 0

### PR DESCRIPTION
**Pull request by javagl**
This addresses https://github.com/CesiumGS/cesium-native/issues/30 (but of course, it's open-ended...).

When running the analysis with the Visual Studio analysis plugin, as described in the linked issue, it initially generated 44877 warnings. No, I'm not trembling due to a caffeine overdose: This is not 487, but 44877. But of course, many (most) of these are in "external" projects (and for some reason, the filtering mechanism to exclude these does not work as desired). And many of them are duplicates, in that they are reported multiple times, e.g. when there is an include of a templated class.

This PR addresses the following warnings:

---

### Exception specifiers

These are mainly missing `noexcept` specifiers, as reported via

- C26440: Function ... can be declared noexcept
- C26455: Default constructor may not throw. Declare it 'noexcept'

This has been a bit of back and forth, because after adding the `noexcept` in the places where it was possible, ... the analyzer found _new_ places where it could _also_ be added. Conversely, there are some places where it _should_ be added, but it was not immediately possible. Some (probably not all) of them are listed here:

- `GltfReader` default constructor calls `registerExtension`, which may throw
- `SchedulerScope` default constructor may throw (if called with an argument - is this still a "default constructor" then?)

- In `Math.h`

        static double lerp(double p, double q, double time) noexcept {
          return glm::mix(p, q, time);
        }

  may throw. Why? Maybe we should just insert our own `mix` there. It's trivial, and the type is pinned to `double` anyhow.

- Tileset destructor calls
   `this->_asyncSystem.dispatchMainThreadTasks();`
  which may throw...
  
(An aside: I think that methods in interfaces should be `noexcept`,   unless there is an *explicit*, documented (!) throws specification)

---

## Missing `const` or `constexpr`

As reported via 

- C26496: The variable ... is assigned only once, mark it as const
- C26814: The const variable ... can be computed at compile-time. Consider using constexpr

This may be the least controversial one. Just add them. It cannot harm.

<sup>Although I already see myself replacing some `const auto` with `auto const`, for sake of "consistency")</sup>

---

## Uninitialized members

As reported via

- C26495: Variable ... is uninitialized. Always initialize a member variable.

Initialized them (everything else could lead to UB, to my understanding - particularly for pointers). Sometimes hesitating about the "most sensible initial value", but usually used `value{}`, with a few cases of `p=nullptr`, `i=0` or `i=-1` for clarity.

---

## Unscoped enum

As reported vie 

- C26812: The enum type 'CesiumGeometry::Axis' is unscoped. Prefer 'enum class' over 'enum'

(Only this case)

---

## Unnecessary copy

As reported via 

- C26817: Potentially expensive copy of variable 'coverageArea' in range-for loop. Consider making it a const reference (es.71).

(Only this case)

---

## Wrong parameter type

As reported via

- C6330: 'char' passed as _Param_(1) when 'unsigned char' is required in call to 'isspace'.

(Only one place)

---

## Reference parameters that can be const

As reported via

- C26460: The reference argument ... for function ... can be marked as const

Fairly uncontroversial, too...

---

# TODO:

Unrelated: The implementation of `RasterizedPolygonsTileExcluder` `getOverlay` is not found

There are still 39941 warnings left, with many of them in our code, and some of them are "critical". For example, there are further functions that are declared `noexcept`, but call throwing functions. It's not always clear how to fix this.

Many other warnings are related to casts or certain security checks (e.g. `nullptr`-checks, or indexed access checks), and I'll address them dedicately in a second pass (and maybe more fine-grained than this PR). 

There are further warnings, of which I don't exactly know (yet) what they mean, even less have an idea how to fix them. Among them are "beautiful" ones like 

> 45>C:\...\cesium-native\Cesium3DTilesSelection\src\TileContentFactory.cpp(97): warning C26426: Global initializer calls a non-constexpr function 'std::unordered_map<std::basic_string<char,std::char_traits<char>,std::allocator<char> >,std::shared_ptr<Cesium3DTilesSelection::TileContentLoader>,std::hash<std::basic_string<char,std::char_traits<char>,std::allocator<char> > >,std::equal_to<std::basic_string<char,std::char_traits<char>,std::allocator<char> > >,std::allocator<std::pair<std::basic_string<char,std::char_traits<char>,std::allocator<char> > const ,std::shared_ptr<Cesium3DTilesSelection::TileContentLoader> > > >::unordered_map<std::basic_string<char,std::char_traits<char>,std::allocator<char> >,std::shared_ptr<Cesium3DTilesSelection::TileContentLoader>,std::hash<std::basic_string<char,std::char_traits<char>,std::allocator<char> > >,std::equal_to<std::basic_string<char,std::char_traits<char>,std::allocator<char> > >,std::allocator<std::pair<std::basic_string<char,std::char_traits<char>,std::allocator<char> > const ,std::shared_ptr<Cesium3DTilesSelection::TileContentLoader> > > >' (i.22).

I'll have a closer look at these, and bring them up as necessary.

The commits in this PR should *largely* be organized by the warning class that was addressed (with `noexcept` being spread over multiple commits), maybe somebody wants ot have a short look.

### CODE REVIEW

**Comment by lilleyse** [comment]
> But of course, many (most) of these are in "external" projects (and for some reason, the filtering mechanism to exclude these does not work as desired).

We had a similar problem in `tilers`. This might be relevant: https://cmake.org/cmake/help/git-master/release/dev/msvc-isystem.html

Basically you need CMake 3.22 (not released yet), Visual Studio 2019 (version 16.10), and the Ninja or NMake generator. Only then will `-external:I` work.

**Comment by javagl** [comment]
@lilleyse Thanks for that hint. For now, I used the workaround of filtering the "Error List" in Visual Studio based on the "File" column, which is clumsy, but worked for that first pass. Also, the analysis was enabled only locally, but I already started looking into the other options, to address https://github.com/CesiumGS/cesium-native/issues/203 at some point. 

In that context: In how far is https://github.com/CesiumGS/tilers/blob/main/.clang-tidy derived from https://github.com/CesiumGS/engineering/issues/124 ? I.e. is this somewhat agreed on to be "The Cesium Style" (and thus, to be used for `cesium-native` as well), or is it only used/configured for `tilers`?

File: CesiumIonClient/include/CesiumIonClient/Connection.h [hunk 1]
```diff
@@ -53,7 +53,7 @@ class CESIUMASYNC_API Connection {
       int64_t clientID,
       const std::string& redirectPath,
       const std::vector<std::string>& scopes,
-      std::function<void(const std::string&)>&& openUrlCallback,
+      const std::function<void(const std::string&)>&& openUrlCallback,
```
Comment (kring) [inline CesiumIonClient/include/CesiumIonClient/Connection.h:56]: Another const r-value ref.

**Review by kring — COMMENTED** [review]
Thanks @javagl, looks good! I looked through all the changes and my only concern is about some r-value references that became const.

File: Cesium3DTilesSelection/include/Cesium3DTilesSelection/Tile.h [hunk 2]
```diff
@@ -556,7 +556,8 @@ class CESIUM3DTILESSELECTION_API Tile final {
    * This method should only be called when this tile's parent is already
    * loaded.
    */
-  void upsampleParent(std::vector<CesiumGeospatial::Projection>&& projections);
+  void
+  upsampleParent(const std::vector<CesiumGeospatial::Projection>&& projections);
```
[the remark this answers is not in the document]
Comment (javagl) [reply Cesium3DTilesSelection/include/Cesium3DTilesSelection/Tile.h:560]: All this is fortunately part of a [single commit](https://github.com/CesiumGS/cesium-native/pull/347/commits/6a5e0d3d4886c1f4ee408954faf38cf218e96af7).

The rules seem to be subtle, I'll read a bit in https://stackoverflow.com/questions/4938875/do-rvalue-references-to-const-have-any-use and try out a little. One crucial question might be: Should it be **possible** to move-from inside that method? I'll have to wrap my head around that, and consult things like https://stackoverflow.com/questions/12095048/what-constitutes-a-valid-state-for-a-moved-from-object-in-c11 as well. I'm currently thinking about something **roughly** along the lines of

```
size_t size = someVector.size();
doSomethingWith(someVector);
T element = vector[size-1];
```
with `doSomethingWith(std::vector<T>&& v);`, 

<sub>Imagine some "as far as I know" and "to my understanding" sprinkled in here:</sub>

After the call, the vector could be in the "moved-from"-state, and ... it's not clear (for me) what that means, exactly. Using a `const&&` could prevent that.

[the remark this answers is not in the document]
Comment (kring) [reply Cesium3DTilesSelection/include/Cesium3DTilesSelection/Tile.h:560]: Well it'd have to be more like this, right? (I assumed the last line should say `someVector` too)

```
size_t size = someVector.size();
doSomethingWith(std::move(someVector));
T element = someVector[size-1];
```

And that std::move means that after that line, `someVector` is in a "valid but unspecified state". Therefore it is unspecified whether `size - 1` is a valid index, and if it's not, it's undefined behavior.

> Should it be possible to move-from inside that method?

Yes. Technically you can move from any non-const reference. But the r-value reference services two purposes:
* Tells clients to expect the object they pass in to be moved out.
* Requires a temporary or an explicitly std::move'd value to be passed in.

[the remark this answers is not in the document]
Comment (javagl) [reply Cesium3DTilesSelection/include/Cesium3DTilesSelection/Tile.h:560]: I had another short look at this. 

>  I don't think it'll prevent code that attempts to move from compiling, it just won't actually result in a move operation. 

Attempting to do a `std::move(constRvalue)` will issue a [warning](https://docs.microsoft.com/en-us/cpp/code-quality/c26478), because it will actually create a copy.

So except for the ambiguity resolution or its use in return types that are mentioned in the SO answers, it only appears to serve the same purpose as `const` on normal reference parameters. For example, having a method that expects a non-const reference like this
```
void example(std::vector<CesiumGeospatial::Projection>& projections) {}

void Tile::upsampleParent(const std::vector<CesiumGeospatial::Projection>&& projections) {
  example(projections);
  ...
}
```
will not compile. Without the `const`, it will compile.

> > Should it be possible to move-from inside that method?
>
> Yes. 

Is there any reason to allow move-from or other mutating operations on the argument? In some way, I'd argue along the line of "make `const` what can be `const`, unless there is a reason do dedicatedly _not_ do this". On the other hand, in the cases that we're talking about here, the semantics are basically: "You have to `move` an object into that function, so the function is fully responsible for that object anyhow".

I think that there is no _really_ strong case for or against `const` here, so I'll leave the decision up to you (i.e. just say "keep the `const`s" or "remove the `const`s", and I'll update the PR accordingly).

[the remark this answers is not in the document]
Comment (kring) [reply Cesium3DTilesSelection/include/Cesium3DTilesSelection/Tile.h:590]: > Attempting to do a std::move(constRvalue) will issue a warning, because it will actually create a copy.

That's a C++ Core Guidelines checker warning, though, righy? Not a regular compiler warning. So it won't (currently) show up in CI, for example.

> Is there any reason to allow move-from or other mutating operations on the argument?

In this case, yes absolutely. Moving `projections` into `upsampleParent` and then into the worker lambda avoids a heap allocation. That heap allocation isn't the end of the world here, of course, but it's so easy to avoid that there's no reason not to.

So in summary:
1. Don't change any code from `Whatever&&` to `const Whatever&&` because that is weird and confusing at best and totally breaks the intention of the code at worst.
2. Don't change any code from `Whatever&&` to `const Whatever&` in this PR because we're _probably_ using move semantics for a reason and  we shouldn't change that in a PR like this with over hundred files changed. If you think move semantics are unnecessary or harmful somewhere, open it as a separate PR.

I think you may have already done this. Will double check and merge.

**Comment by kring** [comment]
There still one more && that got turned into a const&&, in `Connection.h`. Then this is ready to merge.

**Comment by kring** [comment]
:shipit: :tada:
